### DISCUSSION TITLE
Disabling memory-hungry components. Huge Lib.

### DISCUSSION THREAD

**Discussion by Petrus125** [tttapa/Control-Surface#838] [Q&A]
"Sketch uses 21404 bytes (74%) of program storage space. Maximum is 28672 bytes.
Global variables use 2077 bytes (81%) of dynamic memory, leaving 483 bytes for local variables. Maximum is 2560 bytes.
Low memory available, stability problems may occur."

Using Leonardo, Micro, etc, with really small projects, I get this scary message. The library is really huge, so very low space for resources.
Is is posible to reduce/disable-not-used/ parts frm the library, and so have more storage/memory for the sketch?

FAQS has some tips " Reducing the number of individual MIDI elements""Disabling memory-hungry components", but still needmore. Any tip?

**Comment by tttapa** [comment]
You could try using the [`disable-pipes`](https://github.com/tttapa/Control-Surface/tree/disable-pipes) branch and setting `DISABLE_PIPES` to `1`:

https://github.com/tttapa/Control-Surface/blob/91b605873905a6fcb78324052c97dbac10849539/src/Settings/Settings.hpp#L63-L66

Beyond that, I would recommend simply getting a board with more RAM. Much more powerful and feature rich Arduino-compatible boards with ARM and Espressif chips are more suitable for projects like this.

---

If you want to keep using an AVR board with limited memory, you'll have to manually disable parts of Control Surface, or write your own version. It can certainly be done, but then you loose some of Control Surface's high-level features that make it easy to use. 

You can inspect the memory usage using `nm`. Replace `XXXXX` and `YYYYY` by the names printed by the IDE in verbose compilation mode:
```sh
~/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino7/bin/avr-nm -CS --size-sort /tmp/arduino-sketch-XXXXX/YYYYY.ino.elf | egrep '[0-9a-f]{8} [BbDd] '
```
This lists static/global RAM usage, the second column contains the size. As you can see, a lot of the memory is used by MIDI and Serial buffers and by Control Surface vtables:
```cpp
008003c1 00000001 b RxLEDPulse
008003e8 00000001 b timer0_fract
008003c2 00000001 b TxLEDPulse
008003c5 00000001 b _usbConfiguration
008003c4 00000001 b _usbCurrentStatus
008003c3 00000001 b _usbSuspendState
008003c0 00000001 b wdtcsr_save
008001cf 00000001 d CS::CABLE_1
00800592 00000002 B __brkval
00800594 00000002 B __flp
0080011d 00000002 D __malloc_heap_end
0080011f 00000002 D __malloc_heap_start
00800121 00000002 D __malloc_margin
008003bc 00000002 b _cend
008003be 00000002 b _cmark
008003c6 00000002 b CS::MIDI_Interface::DefaultMIDI_Interface
00800495 00000002 b CS::Control_Surface
008003e9 00000004 b timer0_millis
008003ed 00000004 b timer0_overflow_count
00800100 00000004 d breakValue
008003e0 00000004 b AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)144> >::updatables
008003dc 00000004 b AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)160> >::updatables
008003d8 00000004 b AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)176> >::updatables
008003d4 00000004 b AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)192> >::updatables
008003d0 00000004 b AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)208> >::updatables
008003cc 00000004 b AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)224> >::updatables
008003c8 00000004 b AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)240> >::updatables
008003e4 00000004 b AH::UpdatableCRTP<AH::ExtendedIOElement>::updatables
0080049b 00000004 b AH::UpdatableCRTP<AH::Updatable<CS::MIDI_Interface> >::updatables
00800497 00000004 b AH::UpdatableCRTP<AH::Updatable<AH::NormalUpdatable> >::updatables
008003b0 00000004 b PluggableUSB()::obj
00800116 00000007 d _initEndpoints
008003b4 00000008 b guard variable for PluggableUSB()::obj
0080048d 00000008 b guard variable for CS::Control_Surface_::getInstance()::instance
0080010e 00000008 d _usbLineInfo
008001d0 00000008 d vtable for AH::UpdatableCRTP<AH::Updatable<CS::MIDI_Interface> >
008001bb 00000008 d vtable for AH::UpdatableCRTP<AH::Updatable<AH::NormalUpdatable> >
00800104 0000000a d intFunc
00800441 0000000c b MidiUSB
00800159 0000000c d vtable for MIDI_
008001d8 0000000c d vtable for AH::Updatable<CS::MIDI_Interface>
008001d8 0000000c d vtable for AH::Updatable<AH::NormalUpdatable>
008001c3 0000000c d vtable for CS::CCPotentiometer
00800123 0000000e d vtable for CS::MIDI_Source
0080049f 00000011 b potentiometer
00800147 00000012 d vtable for Serial_
00800131 00000016 d vtable for CS::MIDI_Sink
00800197 00000024 d vtable for CS::Control_Surface_
0080022e 00000024 d vtable for CS::TrueMIDI_SinkSource
00800165 00000032 d vtable for CS::MIDI_Pipe
0080044d 00000040 b CS::Control_Surface_::getInstance()::instance
008001e4 0000004a d vtable for CS::MIDI_Interface
00800252 0000004c d vtable for CS::USBMIDI_Interface
008003f1 00000050 b Serial
008004b0 000000e2 b midi
008002a8 00000108 b midi_rx_buffer
```

**Reply by mwpmorris** [reply]
> You could try using the [`disable-pipes`](https://github.com/tttapa/Control-Surface/tree/disable-pipes) branch and setting `DISABLE_PIPES` to `1`:

Hi, I would also like to save some space like the OP but am having difficulty compiling with the "disable-pipes" branch:

Small sample of compilation errors below:
`(error: 'Encoder' does not name a type)`
`error: no matching function for call to 'CS::BorrowedCCRotaryEncoder::BorrowedCCRotaryEncoder`
`/Arduino/libraries/Control-Surface-disable-pipes/src/MIDI_Outputs/CCRotaryEncoder.hpp:62:7: note:   candidate expects 1 argument, 4 provided`

Everything compiles fine with the master branch. Any ideas? Many thanks.

**Comment by Petrus125** [comment]
Tried to reduce resources to minimum, but obviously the size of libs is huge. Very little can be done, for Leonardos and Micros. As mentioned, these boards are probably the best choice for small-medium projects, as they are based on 32U4 ClassCompliant, so ideal for Ipad midi projects.

As you can see, this shown sketch is now over 61%. Anything added, easily reach 80%, as my case.


<img width="923" alt="Foto" src="https://user-images.githubusercontent.com/112938746/200801123-3b388352-ba24-4006-85ff-745b7e97c3fb.png">

**Reply by tttapa** [reply]
> As you can see, this shown sketch is now over 61%. Anything added, easily reach 80%, as my case.

If you look at the memory usage, you get:

```sh
# Internal Arduino (118 bytes)
0x00000001 b "RxLEDPulse"
0x00000001 b "timer0_fract"
0x00000001 b "TxLEDPulse"
0x00000001 b "_usbConfiguration"
0x00000001 b "_usbCurrentStatus"
0x00000001 b "_usbSuspendState"
0x00000001 b "wdtcsr_save"
0x00000002 B "__brkval"
0x00000002 B "__flp"
0x00000002 D "__malloc_heap_end"
0x00000002 D "__malloc_heap_start"
0x00000002 D "__malloc_margin"
0x00000002 b "_cend"
0x00000002 b "_cmark"
0x00000004 b "timer0_millis"
0x00000004 b "timer0_overflow_count"
0x00000004 d "breakValue"
0x00000004 b "PluggableUSB()::obj"
0x00000007 d "_initEndpoints"
0x00000008 d "_usbLineInfo"
0x0000000a d "intFunc"
0x00000008 b "guard variable for PluggableUSB()::obj"
0x0000000c d "vtable for MIDI_"
0x00000012 d "vtable for HardwareSerial"
0x00000012 d "vtable for Serial_"

# Internal Control Surface bookkeeping (739 bytes)
0x00000001 d "CS::CABLE_1"
0x00000002 b "CS::MIDI_Interface::DefaultMIDI_Interface"
0x00000004 b "AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)144> >::updatables"
0x00000004 b "AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)160> >::updatables"
0x00000004 b "AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)176> >::updatables"
0x00000004 b "AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)192> >::updatables"
0x00000004 b "AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)208> >::updatables"
0x00000004 b "AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)224> >::updatables"
0x00000004 b "AH::UpdatableCRTP<CS::MIDIInputElement<(CS::MIDIMessageType)240> >::updatables"
0x00000004 b "AH::UpdatableCRTP<AH::Updatable<CS::MIDI_Interface> >::updatables"
0x00000008 b "guard variable for CS::Control_Surface_::getInstance()::instance"
0x00000008 d "vtable for AH::UpdatableCRTP<AH::Updatable<CS::MIDI_Interface> >"
0x0000000c d "vtable for AH::Updatable<CS::MIDI_Interface>"
0x0000000e d "vtable for CS::MIDI_Source"
0x00000016 d "vtable for CS::MIDI_Sink"
0x00000024 d "vtable for CS::Control_Surface_"
0x00000024 d "vtable for CS::TrueMIDI_SinkSource"
0x00000032 d "vtable for CS::MIDI_Pipe"
0x00000040 b "CS::Control_Surface_::getInstance()::instance"
0x0000004a d "vtable for CS::MIDI_Interface"
0x0000004c d "vtable for CS::USBMIDI_Interface"
0x0000004c d "vtable for CS::SerialMIDI_Interface<HardwareSerial>"
0x0000004c d "vtable for CS::StreamMIDI_Interface"
0x0000004c d "vtable for CS::GenericUSBMIDI_Interface<CS::MIDIUSB_USBDeviceMIDIBackend>"
0x0000004c d "vtable for CS::HardwareSerialMIDI_Interface"

# Instances used in the sketch (905)
0x0000000c b "MidiUSB"
0x00000050 b "Serial"
0x0000009d b "Serial1"
0x000000a6 b "midiser"
0x000000e2 b "midiusb"
0x00000108 b "midi_rx_buffer"
```
So you lose around 118 bytes of RAM to internal Arduino variables and vtables, 739 bytes for internal bookkeeping by Control Surface, and 905 bytes to the Serial and MIDI interfaces you're using in your sketch.  

Most of the memory is used for buffers: a 128 bytes SysEx buffer for each MIDI interface, >64 bytes for the Serial interface, >150 bytes Serial 1 (mostly for the RX and TX buffers), >256 bytes for the MIDI USB buffer.  
You can play with the buffer sizes if you want, e.g. Control_Surface allows you to set the size of the SysEx buffers in Settings.hpp, and you can even disable SysEx input entirely if you don't need it.

The memory overhead by Control Surface comes in two main forms: updatables and vtables.  
Updatables keep track of all the objects instantiated by the user, so the library can automatically discover them, initialize them, and call their `update()` methods automagically when you call `Control_Surface.loop()`, etc.  
Vtables (virtual function tables) implement run-time polymorphism. This is what makes it possible to have many different types (the different MIDI interfaces, the different MIDI Control Elements, etc.) to be used interchangeably, with a homogeneous interface to the user.  
They do add some overhead, but they are at the core of what Control Surface aims to be: a user-friendly library that allows you to create a fully working MIDI controller in a handful lines of code, while also being modular and extensible when more customization is required.

You could remove some of that overhead, but then the resulting library wouldn't be “Control Surface” (or even “Control Surface Lite”), and would require more work from the user.
Perhaps some optimizations can be done (e.g. minimizing the number of virtual functions, in favor of enums and switch statements), but the returns will be minor, and this is not something I can afford to invest time in at the moment.

> but for most common projects, Micro and ProMicro are the choices ( for size, price, specs, etc.)

An Arduino Micro with an 8-bit AVR MCU and 2.5 KiB RAM costs $24.90. In contrast, a Raspberry Pi Pico with a 32-bit ARM Cortex-M0+, 264KiB of RAM and 2 MiB of flash costs just $4.00. A Teensy LC costs $11.65, less than half the price of the Micro, with vastly superior specs.  
It just doesn't make a whole lot of sense to try to fit large MIDI controller projects onto an outdated AVR MCU if you have these cheaper, much more powerful alternatives in the same form factor.

Optimizing every last byte of RAM usage in Control Surface would take hundreds or thousands of dollars in engineering hours, and the result would probably be worse and harder to use for the end user. On the other hand, using a more powerful board will actually save you money, without compromising on user experience.

**Comment by Petrus125** [comment]
>It just doesn't make a whole lot of sense to try to fit large MIDI controller projects onto an outdated AVR MCU if you have these cheaper, much more powerful alternatives in the same form factor.<

Absolutely agree 1000% your point of view. A Lite version, would be kind of going back " to XP compatible". Makes no sense.Thx.
